source/funkin/util/macro/RegistryMacro.hx

```diff
@@ -0,0 +1,357 @@
+    fields = fields.concat(buildEntryMethods(cls));
+
+    cls.meta.add(":funkinProcessed", [], cls.pos);
+
+    return fields;
+  }
+
+  #if macro
+  static function fieldAlreadyExists(name:String):Bool
+  {
+    for (field in Context.getBuildFields())
+    {
+      if (field.name == name && !((field.access ?? []).contains(Access.AAbstract)))
+      {
+        return true;
+      }
+    }
+
+    function fieldAlreadyExistsSuper(name:String, superClass:Null<ClassType>)
+    {
+      if (superClass == null)
+      {
+        return false;
+      }
+
+      for (field in superClass.fields.get())
+      {
+        if (field.name == name && !field.isAbstract)
+        {
+          return true;
+        }
+      }
+
+      // recursively check superclasses
+      return fieldAlreadyExistsSuper(name, superClass.superClass?.t.get());
+    }
+
+    return fieldAlreadyExistsSuper(name, Context.getLocalClass().get().superClass?.t.get());
+  }
+
+  static function alreadyProcessed(cls:ClassType):Bool
+  {
+    var processedMeta = cls.meta.get().find(function(m) return m.name == ':funkinProcessed');
+
+    if (processedMeta != null)
+    {
+      return true;
+    }
+
+    if (cls.superClass != null)
+    {
+      return alreadyProcessed(cls.superClass.t.get());
+    }
+
+    return false;
+  }
+
+  static function getTypeParams(cls:ClassType):RegistryTypeParamsNew
+  {
+    switch (cls.superClass.t.get().kind)
+    {
+      case KGenericInstance(_, params):
+        var typeParams:Array<Dynamic> = [];
+        for (param in params)
+        {
+          switch (param)
+          {
+            case TInst(t, _):
+              typeParams.push(t.get());
+            case TType(t, _):
+              typeParams.push(t.get());
+            default:
+              throw 'Not a class';
+          }
+        }
+        return {entryCls: typeParams[0], jsonCls: typeParams[1]};
+      default:
+        throw 'Not in the correct format';
+    }
+  }
+
+  static function getEntryData(cls:ClassType):Dynamic // DefType or ClassType
+  {
+    try
+    {
+      switch (cls.interfaces[0].params[0])
+      {
+        case Type.TInst(t, _):
+          return t.get();
+        case Type.TType(t, _):
+          return t.get();
+        default:
+          throw 'Entry Data is not a class or typedef';
+      }
+    }
+    catch (e)
+    {
+      throw '
+        ${cls.name}:
+        Make sure IRegistryEntry is the last implement
+        class ExampleEntry implements ... implements IRegistryEntry
+      ';
+    }
+  }
+
+  static function buildRegistryVariables(cls:ClassType):Array<Field>
+  {
+    var clsType:ComplexType = Context.getType('${cls.module}.${cls.name}').toComplexType();
+
+    var newInstance:String = 'new ${cls.module}.${cls.name}()';
+
+    return (macro class TempClass
+      {
+        static var _instance:Null<$clsType>;
+        public static var instance(get, never):$clsType;
+
+        static function get_instance():$clsType
```

Comment:
This would be nice to migrate into its own `ISingleton` macro and have everything with an `instance` variable use it.